### PR TITLE
Remove Hardhat

### PR DESCRIPTION
**Pull request by Bridgerz**

### CODE REVIEW

File: contracts/utils/MessageVerifier.sol [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+abstract contract MessageVerifier is Initializable {
+    IBridgeCommittee public committee;
+    // messageType => nonce
+    mapping(uint8 => uint64) public nonces;
+
+    function __MessageVerifier_init(address _committee) internal onlyInitializing {
+        committee = IBridgeCommittee(_committee);
+    }
+
+    modifier verifySignatures(
```
Comment (longbowlu) [inline contracts/utils/MessageVerifier.sol]: -> `verifySignaturesAndNonce`

File: contracts/utils/CommitteeUpgradeable.sol [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        __ReentrancyGuard_init();
+        __MessageVerifier_init(_committee);
+        committee = IBridgeCommittee(_committee);
+    }
+
+    function _authorizeUpgrade(address) internal view override {
+        require(_upgradeAuthorized, "SuiBridge: Unauthorized upgrade");
+    }
+
+    function upgradeWithSignatures(bytes[] memory signatures, BridgeMessage.Message memory message)
```
Comment (longbowlu) [inline contracts/utils/CommitteeUpgradeable.sol:28]: I think this does not work - different contracts need to have separate nonces for upgrade, right?

File: contracts/BridgeTokens.sol [hunk 3]
```diff
@@ -12,7 +12,7 @@ contract BridgeTokens is Ownable, IBridgeTokens {
 
     /// @dev Constructor function
     /// @param _supportedTokens An array of addresses representing the supported tokens
-    constructor(address[] memory _supportedTokens) Ownable() {
+    constructor(address[] memory _supportedTokens) Ownable(msg.sender) {
```
Comment (longbowlu) [inline contracts/BridgeTokens.sol:15]: why this change?

File: contracts/utils/CommitteeUpgradeable.sol [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        verifySignatures(message, signatures, BridgeMessage.BRIDGE_UPGRADE)
+    {
+        // decode the upgrade payload
+        (address implementationAddress, bytes memory callData) =
+            BridgeMessage.decodeUpgradePayload(message.payload);
+
+        // authorize upgrade
+        _upgradeAuthorized = true;
+        // upgrade contract
+        upgradeToAndCall(implementationAddress, callData);
```
Comment (longbowlu) [inline contracts/utils/CommitteeUpgradeable.sol:40]: did you have a chance to check if this is reentrance free?

File: test/BridgeBaseTest.t.sol [hunk 5]
```diff
@@ -48,65 +48,64 @@ contract BridgeBaseTest is Test {
...
-
-        vm.expectRevert(bytes("BridgeCommittee: Total stake must be 10000"));
-        committee.initialize(_committee, _stake);
-    }
-*/
-
+    /**
+     * // TODO: It does not run if it is outside of setUpBridgeTest
+     * function testFailInitializeCommitteeAndStakeArraysMustBeOfTheSameLength() public {
+     *     address[] memory _committee = new address[](5);
```
Comment (longbowlu) [inline test/BridgeBaseTest.t.sol:54]: what's the plan to fix these?

File: contracts/BridgeTokens.sol (hunk 3 above)
Comment (Bridgerz) [reply contracts/BridgeTokens.sol:15]: Using OZ V5 library now to support the new OZ upgrades library. There are a few minor interface changes that were needed like this.

File: contracts/utils/CommitteeUpgradeable.sol (hunk 4 above)
Comment (Bridgerz) [reply contracts/utils/CommitteeUpgradeable.sol:40]: upgradeToAndCall is not itself nonReentrant but all "WithSignature" (ex. `upgradeWithSignatures`) functions should be.

File: test/BridgeBaseTest.t.sol (hunk 5 above)
Comment (Bridgerz) [reply test/BridgeBaseTest.t.sol:54]: These are the fuzzing tests I was talking about. @georgedigkas should be on this?

File: contracts/utils/CommitteeUpgradeable.sol (hunk 2 above)
Comment (Bridgerz) [reply contracts/utils/CommitteeUpgradeable.sol:28]: Screenshot 2024-01-26 at 12 35 38 PM

File: contracts/BridgeLimiter.sol [hunk 6]
```diff
@@ -66,13 +54,18 @@ contract BridgeLimiter is
         return windowAmount + USDAmount > totalLimit;
     }
 
+    function willUSDAmountExceedLimit(uint256 amount) public view returns (bool) {
+        uint256 windowAmount = calculateWindowAmount();
+        return windowAmount + amount > totalLimit;
+    }
+
```
Comment (longbowlu) [inline contracts/BridgeLimiter.sol:64]: nit: we can change to `willAmountExceedLimit`
```
    function willAmountExceedLimit(uint8 tokenId, uint256 amount) public view returns (bool) {
        uint256 USDAmount = calculateAmountInUSD(tokenId, amount);
        return willUSDAmountExceedLimit(USDAmount);
    }
```

Comment (Bridgerz) [reply contracts/BridgeLimiter.sol:64]: Remember there are two functions. `willAmountExceedLimit(uint8 tokenId, uint256 amount)` and `function willUSDAmountExceedLimit(uint256 amount)` both used for different purposes.

File: contracts/utils/CommitteeUpgradeable.sol (hunk 4 above)
Comment (longbowlu) [reply contracts/utils/CommitteeUpgradeable.sol:40]: so the theory is as long as the callData (which is signed by validators) is not reentrant-able, this is safe?

Comment (Bridgerz) [reply contracts/utils/CommitteeUpgradeable.sol:40]: correct

**Review by longbowlu — APPROVED** [review]
